File: protocol/csharp/ToInfinity.Protocol/ControlFrame.cs
using System.Buffers.Binary;
using System.Text;

namespace ToInfinity.Protocol;

/// <summary>
/// SPEC.md §4 frame kind tag distinguishing JSON control messages from
/// binary input events on the shared control-channel connection.
/// </summary>
public enum ControlFrameKind : byte
{
    Json = 0x01,
    InputEvent = 0x02,
}

/// <summary>
/// Length-prefixed framing for the control channel (SPEC.md §2, amended by
/// §4): every frame is <c>[len:4 big-endian][kind:1][payload:len-1]</c>.
/// This class only handles the framing envelope; JSON body encode/decode is
/// <see cref="ControlMessage"/> and binary event body encode/decode is
/// <see cref="InputEvent"/>.
/// </summary>
public static class ControlFrame
{
    /// <summary>Size in bytes of the leading length prefix.</summary>
    public const int LengthPrefixSize = 4;

    /// <summary>Size in bytes of the frame-kind tag that follows the length prefix.</summary>
    public const int KindTagSize = 1;

    /// <summary>
    /// Encodes a full frame (length prefix + kind tag + payload) for the given kind and payload bytes.
    /// </summary>
    public static byte[] Encode(ControlFrameKind kind, ReadOnlySpan<byte> payload)
    {
        int bodyLength = KindTagSize + payload.Length; // length field counts kind byte + payload
        var buffer = new byte[LengthPrefixSize + bodyLength];
        BinaryPrimitives.WriteUInt32BigEndian(buffer.AsSpan(0, 4), (uint)bodyLength);
        buffer[4] = (byte)kind;
        payload.CopyTo(buffer.AsSpan(5));
        return buffer;
    }

    /// <summary>Encodes a JSON control message as a full framed byte array.</summary>
    public static byte[] EncodeJson(ControlMessage message)
        => Encode(ControlFrameKind.Json, Encoding.UTF8.GetBytes(message.ToJson()));

    /// <summary>Encodes a binary input event as a full framed byte array.</summary>
    public static byte[] EncodeInputEvent(InputEvent inputEvent)
        => Encode(ControlFrameKind.InputEvent, inputEvent.Encode());

    /// <summary>
    /// Reads the 4-byte big-endian length prefix. The returned value is the
    /// byte length of everything after the prefix (kind tag + payload).
    /// Throws <see cref="ArgumentException"/> if the declared length exceeds
    /// <see cref="ProtocolConstants.MaxControlFrameSize"/>.
    /// </summary>
    public static int ReadLengthPrefix(ReadOnlySpan<byte> fourBytes)
    {
        if (fourBytes.Length != LengthPrefixSize)
        {
            throw new ArgumentException($"Length prefix must be exactly {LengthPrefixSize} bytes.", nameof(fourBytes));
        }

        uint length = BinaryPrimitives.ReadUInt32BigEndian(fourBytes);
        if (length > ProtocolConstants.MaxControlFrameSize)
        {
            throw new ArgumentException(
                $"Declared frame length {length} exceeds max {ProtocolConstants.MaxControlFrameSize}.");
        }

        return (int)length;
    }

    /// <summary>Result of decoding a frame body (the bytes after the length prefix).</summary>
    public readonly struct DecodedFrame
    {
        public DecodedFrame(ControlFrameKind kind, byte[] payload)
        {
            Kind = kind;
            Payload = payload;
        }

        public ControlFrameKind Kind { get; }
        public byte[] Payload { get; }
    }

    /// <summary>
    /// Decodes a frame body — the bytes immediately following the 4-byte
    /// length prefix, i.e. <c>[kind:1][payload:...]</c> — into its kind tag
    /// and payload bytes. Throws <see cref="ArgumentException"/> if the body
    /// is empty (missing kind tag).
    /// </summary>
    public static DecodedFrame DecodeBody(ReadOnlySpan<byte> body)
    {
        if (body.Length < KindTagSize)
        {
            throw new ArgumentException("Frame body is too short to contain a kind tag.", nameof(body));
        }

        var kind = (ControlFrameKind)body[0];
        var payload = body[1..].ToArray();
        return new DecodedFrame(kind, payload);
    }
}

File: protocol/swift/ToInfinityProtocol/Sources/ToInfinityProtocol/ControlFrame.swift
import Foundation

/// SPEC.md §4 frame kind tag distinguishing JSON control messages from
/// binary input events on the shared control-channel connection. Modeled
/// as a raw-byte wrapper (not a closed `enum`) to mirror the C#
/// `ControlFrameKind` enum's permissive byte cast semantics.
public struct ControlFrameKind: RawRepresentable, Equatable {
    public let rawValue: UInt8

    public init(rawValue: UInt8) {
        self.rawValue = rawValue
    }

    public static let json = ControlFrameKind(rawValue: 0x01)
    public static let inputEvent = ControlFrameKind(rawValue: 0x02)
}

/// Length-prefixed framing for the control channel (SPEC.md §2, amended by
/// §4): every frame is `[len:4 big-endian][kind:1][payload:len-1]`. This
/// type only handles the framing envelope; JSON body encode/decode is
/// `ControlMessageCodec` and binary event body encode/decode is `InputEvent`.
public enum ControlFrame {
    /// Size in bytes of the leading length prefix.
    public static let lengthPrefixSize = 4
    /// Size in bytes of the frame-kind tag that follows the length prefix.
    public static let kindTagSize = 1

    /// Encodes a full frame (length prefix + kind tag + payload) for the given kind and payload bytes.
    public static func encode(kind: ControlFrameKind, payload: Data) -> Data {
        let bodyLength = kindTagSize + payload.count
        var bytes = BigEndian.writeUInt32(UInt32(bodyLength))
        bytes.append(kind.rawValue)
        var data = Data(bytes)
        data.append(payload)
        return data
    }

    /// Encodes a JSON control message as a full framed byte array.
    public static func encodeJson(_ message: ControlMessage) throws -> Data {
        let json = try ControlMessageCodec.encode(message)
        return encode(kind: .json, payload: json)
    }

    /// Encodes a binary input event as a full framed byte array.
    public static func encodeInputEvent(_ event: InputEvent) -> Data {
        encode(kind: .inputEvent, payload: event.encode())
    }

    /// Reads the 4-byte big-endian length prefix. The returned value is the
    /// byte length of everything after the prefix (kind tag + payload).
    /// Throws `ProtocolCodecError.frameTooLarge` if the declared length
    /// exceeds `ProtocolConstants.maxControlFrameSize`, or
    /// `ProtocolCodecError.malformedFrame` if `fourBytes` is not exactly 4 bytes.
    public static func readLengthPrefix(_ fourBytes: Data) throws -> Int {
        guard fourBytes.count == lengthPrefixSize else {
            throw ProtocolCodecError.malformedFrame
        }

        let length = BigEndian.readUInt32([UInt8](fourBytes), at: 0)
        guard length <= UInt32(ProtocolConstants.maxControlFrameSize) else {
            throw ProtocolCodecError.frameTooLarge
        }

        return Int(length)
    }

    /// Result of decoding a frame body (the bytes after the length prefix).
    public struct DecodedFrame: Equatable {
        public let kind: ControlFrameKind
        public let payload: Data

        public init(kind: ControlFrameKind, payload: Data) {
            self.kind = kind
            self.payload = payload
        }
    }

    /// Decodes a frame body — the bytes immediately following the 4-byte
    /// length prefix, i.e. `[kind:1][payload:...]` — into its kind tag and
    /// payload bytes. Throws `ProtocolCodecError.malformedFrame` if the
    /// body is empty (missing kind tag).
    public static func decodeBody(_ body: Data) throws -> DecodedFrame {
        guard body.count >= kindTagSize else {
            throw ProtocolCodecError.malformedFrame
        }

        let bytes = [UInt8](body)
        let kind = ControlFrameKind(rawValue: bytes[0])
        let payload = Data(bytes[1...])
        return DecodedFrame(kind: kind, payload: payload)
    }
}
